File: Algorithms/dynamic-programming/floyd-warshall.cpp
/* floyd warshall algorithm in cpp: */

#include "iostream"

using namespace std;

#define nV 4
#define INF 999

void printMatrix(int graph[][nV]);

void floydWarshall(int graph[][nV]){
    int resultMatrix[nV][nV], i, j, k;

    /* copy the matrix*/
    for (int i = 0; i < nV; i ++){
        for (int j = 0; j < nV; j ++){
            resultMatrix[i][j] = graph[i][j];
        }
    }

    /* adding vertices individually: */

    for(k = 0; k < nV; k++){
        for (i = 0; i < nV; i ++){
            for (j = 0; j < nV; j ++){
                if (resultMatrix[i][k] + resultMatrix[k][j] < resultMatrix[i][j]){
                    resultMatrix[i][j] = resultMatrix[i][k] + resultMatrix[k][j];
                }
            }
        }
    }

    printMatrix(resultMatrix);
}


/* utility function to print the result matrix: */

void printMatrix(int graph[][nV]){
    for (int i = 0; i < nV; i ++){
        for (int j = 0; j < nV; j ++){
            if( graph[i][j] == INF){
                cout << "INF" << " ";
            } else{
                cout << graph[i][j] << " ";
            }
        }
        cout << endl;
    }
}
/* driver code: */

int main(){
    int graph[nV][nV] = {{0, 3, INF, 5},
                        {2, 0, INF, 4},
                        {INF, 1, 0, INF},
                        {INF, INF, 2, 0}};
  floydWarshall(graph);
}

File: Algorithms/dynamic-programming/floyd-warshall.py
""" Floyd warshall algorithm in python: """

INF = 999

def floydWarshall(G, nV):
    distance = list(map(lambda i : list(map(lambda j : j, i)), G))
    """ adding vertices individually: """
    for k in range(nV):
        for i in range(nV):
            for j in range(nV):
                distance[i][j] = min(distance[i][j], distance[i][k] + distance[k][j])
    printSolution(distance, nV)

def printSolution(distance, nV):
    for i in range(nV):
        for j in range(nV):
            if (distance[i][j] == INF):
                print("INF", end=" ")
            else:
                print(distance[i][j], end= " ")
        print(" ")

if __name__ == "__main__":
    G = [[0, 3, INF, 5],
         [2, 0, INF, 4],
         [INF, 1, 0, INF],
         [INF, INF, 2, 0]]
    print(len(G[0]))
    floydWarshall(G, len(G[0]))